pubsub/envelope.go

```diff
@@ -30,29 +30,55 @@ type CloudEventsEnvelope struct {
 	Subject         string      `json:"subject"`
 }
 
-// NewCloudEventsEnvelope returns a new CloudEventsEnvelope
+// NewCloudEventsEnvelope returns CloudEventsEnvelope from data or a new one when data content was not
+// already an existing CE. Invoked by either the HTTP or gRPC pug handles where:
+//  id = new UUID
+// 	source = handling app ID
+//  eventType = const DefaultCloudEventType from this package ("com.dapr.event.sent")
+//  subject = diagnostics span context W3C as string
 func NewCloudEventsEnvelope(id, source, eventType, subject string, data []byte) *CloudEventsEnvelope {
-	if eventType == "" {
-		eventType = DefaultCloudEventType
+	var ce CloudEventsEnvelope
+	err := jsoniter.Unmarshal(data, &ce)
+	if err != nil {
+		// data content is not JSON
+		ce = CloudEventsEnvelope{
+			Data:            string(data),
+			DataContentType: "text/plain",
+		}
 	}
-	contentType := ""
 
-	var i interface{}
-	err := jsoniter.Unmarshal(data, &i)
-	if err != nil {
-		i = string(data)
-		contentType = "text/plain"
-	} else {
-		contentType = "application/json"
+	// populate the values where CloudEvent has none
+	if ce.ID == "" {
+		ce.ID = id
+	}
+
+	if ce.Source == "" {
+		ce.Source = source
 	}
 
-	return &CloudEventsEnvelope{
-		ID:              id,
-		Source:          source,
-		Type:            eventType,
-		Data:            i,
-		SpecVersion:     CloudEventsSpecVersion,
-		DataContentType: contentType,
-		Subject:         subject,
+	if ce.Type == "" {
```

Comment:
if the published event is already a cloud event, we should just return it as, no need to modify the properties which are in the original cloud event.

Reply:
This is defensive posture to allow for CE optional fields to be set by the invoking Dapr code (HTTP or gRPC pub handlers). I was also trying to adhere to existing test to ensure I don't break something so if the published cloud event is not including eventType for example some of these would tail.

Reply:
Once we enable this functionality, IMO, as a user, if I gave a cloud event, I would not expect Dapr to modify it, it should just take it and publish. Tests can be modified.

Reply:
Refactored to return the CE as is if provided by the user. Also added CE content tests to make sure we are not altering the user defined values